### ISSUE TITLE
Single file model format?

### ISSUE THREAD

**Issue by kba** [Calamari-OCR/calamari#64]
We're building an engine-agnostic training server. Throw ground truth at it, track training progress, download (intermediary) models to evaluate, that sort of thing.

As a convenience, we'd like to bundle the multiple files that represent a model. Do you have an opinion on how or maybe even a plan/code to do that?

We thought the easiest way would be to zip

* `xyz.ckpt.data-00000-of-00001`
* `xyz.ckpt.index`
* `xyz.ckpt.json`
* `xyz.ckpt.meta`

into a flat zip archive and send it to clients with a media type like `application/vnd.calamari.tf+zip`.

**Comment by ChWick** [comment]
A zip archive is a very simple and valid solution.
I was trying to use tensorflow's 'freeze' function which also would result in a single file, there were however some issues regarding the cudnn-lstm cells when loading such a network. This is why I stuck with the ugly 4-file solution.

**Comment by ChWick** [comment]
See also #44

**Comment by bertsky** [comment]
> A zip archive is a very simple and valid solution.
> I was trying to use tensorflow's 'freeze' function which also would result in a single file, there were however some issues regarding the cudnn-lstm cells when loading such a network. This is why I stuck with the ugly 4-file solution.

Sorry for being a few years late, but I also had this issue with my Keras models (depending on whether I wanted to load LSTM weights on CPU or GPU), so I decided to stuff all the necessary configuration for recompiling the model into the HDF5 dictionaries. It's pretty flexible, you can embed all the metadata there I believe. (It would still be multiple .h5 for multiple checkpoints though. And users would not have the advantage of easily inspecting the model codec, postprocessing etc anymore. But you could then make a dedicated tool for inspection...)
